### PR TITLE
Modelling/Training Data Augmentation & Sampling

### PR DESCRIPTION
**Pull request by justinbt1**
# Training Data Augmentation & Sampling

### Additional Data
Added additional HRCS tagged grant dataset from the NIHR, these grants are tagged with both HRCS RAC and HC codes at the time of application. The dataset used is from https://nihr.opendatasoft.com/api/explore/v2.1/catalog/datasets/nihr-summary-view/exports/parquet?lang=en&timezone=Europe%2FLondon and contains grant title and abstract alongside NIHR grant ids and HRCS tags.

Code for loading the NIHR dataset can be found in src/data_processing/nihr_data.py

Distribution of low level HRCS RAC tags following addition of the NIHR dataset. (post-deduplication)
output

This additional data provides support for some of the less represented classes, however the overall distribution is still very skewed.

### De-duplication
Though we currently drop duplicate grants based on organisation name and grant id, having reviewed the literature and run some tests there are a large number of exact and near duplicate texts in the combined dataset.

This is problematic as it adds no additional information for model training and may cause the model to over fit to specific sub-groups of grants. It also increases the risk of data leaking into the evaluation dataset, preventing robust model performance evaluation by artificially boosting model performance metrics.

To resolve this issue a deduplication step has been added to data_processing.py that:
- Uses Pandas for case insensitive removal of identical texts.
- Uses the minhash and LSH algorithms to efficiently identify and drop near-duplicate texts (uses an estimated Jaccard similarity of 0.6).

This removes 5,405 exact duplicates and 4,808 near duplicates from the dataset.

Near duplicates can be caused by different versions/continuations of grants, or different grants with very similar text. 

A shortened example near duplicate:
```
Cell Therapy for Osteoarthritis The RJAH Orthopaedic Hospital Oswestry has pioneered autologous cell therapy
Cell Therapy for Osteoarthritis (ASCOT) The RJAH Orthopaedic Hospital Oswestry has pioneered autologous cell therapy
```

### Multi-label Stratified Sampling
Replaced simple randomised train/test split with multisample stratified split, this both avoids the potential issue of smaller classes having a test sample size of 0 as they have a much lower probability of being selected, leading to imbalance. This has been shown in the literature to improve classifier performance for multi-label classifier tasks on imbalanced data.

Papers:
Sechidis, K., Tsoumakas, G., & Vlahavas, I. (2011). On the stratification of multi-label data. Machine Learning and Knowledge Discovery in Databases, 145-158. http://lpis.csd.auth.gr/publications/sechidis-ecmlpkdd-2011.pdf

Piotr Szymański, Tomasz Kajdanowicz ; Proceedings of the First International Workshop on Learning with Imbalanced Domains: Theory and Applications, PMLR 74:22-35, 2017. http://proceedings.mlr.press/v74/szyma%C5%84ski17a.html

### Boiler-plate Text Removal
Grants from some organisations have boiler plate text such as 'background:' at the start of their abstract texts, this may lead to fitting on text specific to certain orgs instead of meaningful information. Code to remove some of these has been included in this pull request, but more work is needed! 

### Multi-dataset Support
Instead of having to run `data_processing` and `pre-processing` multiple times for each tag type, they now run once and output preprocessed datasets for HC, RA and RA_top tags. A test, train and value count dataset is generated for each in the `data/preprocessed` directory.

### Makefile Changes
- Moved dataset download into a separate Make rule, so download does not need to occur when modifying and testing data processing modules.
- Train rule now requires a `path` argument for training data directory, to allow train to be run on the different RA, RA_top, HC data directories.
- Added rules for running train on the RA, RA_top and HC data types, these wrap and call the train rule with the correct `path` parameter for that data type.

For example, to run train on the HC tag data, both of the below options will work:
```shell
make train_hc
make train path="data/preprocessed/hc"
```
This allows for some flexibility for dealing with alternative file locations for train runs on different datasets, while providing an easy wrapper for the default case.

Note: At present the tag type to train on needs to be specified both in the config and make, it might be better in the future if we can find a way to move tag type specification to a single location.

### Additional Changes
- Data processing has been split into multiple modules, with separate modules for processing the UKHRA and NIHR datasets, with code to combine them and shared transformations  still in `data_processing.py`. This avoids a massive Python file, and necessarily moves some functions around to allow for NIHR data processing and combination.
- Transformers trainer now prints training metrics every epoch.
- Refactored train.py, moving logic from `if __name__ == __main__` block to a new function and un-nesting the `WeightedTrainer` class.
- Updated `wandb` to log additional configuration parameters on model init.
- Modified `MultiLabelBinarizer` so data is stored in columns with headings, instead of single array column. This allows for easier analysis outside of model training. Updated the `train` function in `train.py` to accommodate these changes by converting the tag columns to a Numpy array.
- Updated `tag_overview.ipynb` to show new class distributions.

### CODE REVIEW

**Comment by ArneRobben** [comment]
@justinbt1  Can this be merged into main?

File: src/data_processing/ukhra_data.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+    Args:
+        year(int): Year of UKHRA study.
+
+    Returns:
+        pd.DataFrame: UKHRA dataframe for given year.
+
+    """
+    ukhra_df = pd.read_excel(
+        f'data/raw/ukhra{year}.xlsx',
```
Comment (ArneRobben) [inline src/data_processing/ukhra_data.py:16]: It might be better to feed this as an argument in case either the directory changes or you call this definition from a different working directory.

File: src/data_processing/nihr_data.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+def read_nihr_dataset():
+    """Reads NIHR dataset and clean columns.
+
+    Returns:
+        pd.DataFrame: Cleaned NIHR dataset.
+
+    """
+    nihr_df = pd.read_parquet(
+        'data/raw/nihr_all.parquet',
```
Comment (ArneRobben) [inline src/data_processing/nihr_data.py:12]: similarly, to the `ukhra_data.py` comment, I would make this into an argument, first of the definition but ultimately controllable from the Makefile via click or Argparse

**Comment by justinbt1** [comment]
> @ArneRobben ready for code review!

Ah you've already started!

File: src/data_processing/data_processing.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    signatures = minhash.transform(df['lower'])
+    lsh = akin.LSH(minhash.permutations, seed=1)
+    lsh.update(signatures, df.index)
+    adj = lsh.adjacency_list(min_jaccard=0.6)
+
+    duplicate_ids = set()
+    for k, v in adj.items():
+        duplicate_ids.update(v)
+
+    deduplicated_ids = set(adj.keys()) - duplicate_ids
```
Comment (ArneRobben) [inline src/data_processing/data_processing.py:70]: I had a look at the workings of `deduplicate` by doing some spot checking. Here is an example of a number of grants that are very similar (have the same title), so I was looking at how these would be deduped. 

```
	OrganisationReference	AllText
7303	HRCS2014_02138	Senior Research and Information Nurse Clinical research, clinical trials, patient and profession...
7315	HRCS2014_02121	Senior Research and Information Nurse The post is based at the University of Birmingham’s Cancer...
7331	HRCS2014_02108	Senior Research and Information Nurse The role of the Senior Research Nurse is: To promote th...
7534	HRCS2014_01436	Senior Research and Information Nurse The role of the Senior Research Nurse is: To promote th...
7535	HRCS2014_01437	Senior Research and Information Nurse The role of the Senior Research Nurse is: To promote th...
7536	HRCS2014_01438	Senior Research and Information Nurse The role of the Senior Research Nurse is: To promote th...
7537	HRCS2014_01439	Senior Research and Information Nurse The role of the Senior Research Nurse is: To promote th...
7538	HRCS2014_01440	Senior Research and Information Nurse The role of the Senior Research Nurse is: o To promote th...
7539	HRCS2014_01441	Senior Research and Information Nurse The role of the Senior Research Nurse is: o To promote th...
7540	HRCS2014_01442	Senior Research and Information Nurse The role of the Senior Research Nurse is: o To promote th...
7541	HRCS2014_01443	Senior Research and Information Nurse The role of the Senior Research Nurse is: To promote th...
7542	HRCS2014_01444	Senior Research and Information Nurse This senior research nursing post is based on the early ph...
7543	HRCS2014_01445	Senior Research and Information Nurse The role of the Senior Research Nurse is: To promote th...
```

Good news is that `duplicate_ids` does a very good job and reckognises the exact duplicates. `duplicated_ids` = `{2, 3, 4, 5, 6, 7, 8, 9, 10, 12}`. So from this spot check the Jaccard distance seems reasonable. 

However, the line `deduplicated_ids = set(adj.keys()) - duplicate_ids` kicks all the duplicates out, I would expect one to be kept.

File: src/train.py [hunk 4]
```diff
@@ -163,6 +198,7 @@ def train(
         save_strategy=config['training_settings']['save_strategy'],
         save_total_limit=config['training_settings']['save_total_limit'],
         output_dir=model_path,
+        logging_strategy='epoch'
```
Comment (ArneRobben) [inline src/train.py:201]: for consistency probably worth adding this to the config

File: src/data_processing/data_processing.py (hunk 3 above)
Comment (justinbt1) [reply src/data_processing/data_processing.py:70]: I think the below code should do what we want:
```python
deduplicated_ids = set()
duplicate_ids = set()
for k, vs in adj.items():
    if k not in duplicate_ids:
        deduplicated_ids.add(k)
    duplicate_ids.update(vs)
```

Comment (justinbt1) [reply src/data_processing/data_processing.py]: I've committed the above changes.

**Comment by justinbt1** [comment]
I've explicitly removed the null MRC studentship abstracts and titles from the dataset.
Title: `'mrc studentship - award title not available in public dataset'`
Abstract: `'mrc studentship - award abstract not available in public dataset'`

File: src/train.py (hunk 4 above)
Comment (justinbt1) [reply src/train.py]: Added logging_strategy to `training_settings` in `train_config.yaml` and `Training Arguments` in `train.py`

File: src/data_processing/data_processing.py (hunk 3 above)
Comment (ArneRobben) [reply src/data_processing/data_processing.py]: I ran that bit of code on the grants above to sense check and it works!

Understood about the slight suboptimal nature of the dropping of nodes but don't think it matters much here (as in, it shouldn't matter which of the duplicates we keep)

File: Makefile [hunk 5]
```diff
@@ -3,29 +3,51 @@ help:
...
-		--train-path "data/preprocessed/train.parquet" \
-		--test-path "data/preprocessed/test.parquet" \
-		--value-counts-path "data/preprocessed/value_counts.json" \
+		--train-path "${path}/train.parquet" \
+		--test-path "${path}/test.parquet" \
+		--value-counts-path "${path}/value_counts.json" \
 		--label-names-path "data/label_names/ukhra_ra.jsonl" \
 		--model-dir "data/model/"
+
+.PHONY: train_ra
```
Comment (ArneRobben) [inline Makefile:43]: These are nice additions!

**Comment by justinbt1** [comment]
@ArneRobben with the recent changes are you happy to approve this pull request?

**Comment by ArneRobben** [comment]
> @ArneRobben with the recent changes are you happy to approve this pull request?

Almost.. There's a few things I'd like to check before.

**Comment by justinbt1** [comment]
I think I've found the issue causing the strange labelling behaviour we see in the metrics! 

The `split_data_frame` function in the `preprocess.py` module uses the Sklearn MultiLabelBinarizer to convert the feature labels into a one-hot-encoded format.

The code in this pull request:
```python
mlb = MultiLabelBinarizer()
y = mlb.fit_transform(df[category])

...

train = pd.DataFrame(y_train, columns=mlb.classes_)
test = pd.DataFrame(y_test, columns=mlb.classes_)
```
The MultiLabelBinarizer orders the one-hot-encoded binary columns based on the order of occurrence of the class labels as it traverses the data set when the fit_transform method is called. However, the .classes_ attribute does not honour the class label order and returns the class labels in alphabetical order instead. 

This means the columns of the dataframe are labeled in the wrong order, though this in itself is not responsible for the mislabeling as the column names are not used in `train.py`.

The issue is that in the `train` function of the `train.py` module there is code that sorts value_counts by key:

```python
# sort value_counts by key
value_counts = {
    k: v for k, v in sorted(value_counts.items(), key=lambda item: item[0])
}
```

**Review by ArneRobben — APPROVED** [review]
